src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -57,15 +57,23 @@ class LedgerImpl(vm: VM, blockchainConfig: BlockchainConfig) extends Ledger with
     val blockExecResult = for {
       //todo why do we require in executeBlockTransactions separate blockchain and blockchain storage
       //todo blockchain storage is part of blockchain
-      execResult <- executeBlockTransactions(block, blockchain, storages, validators.signedTransactionValidator).left.map(e => TxError(e.reason))
+      execResult <- executeBlockTransactions(block, blockchain, storages, validators.signedTransactionValidator)
       BlockResult(resultingWorldStateProxy, _, _) = execResult
       worldToPersist = payBlockReward(blockchainConfig.blockReward, block, resultingWorldStateProxy)
       worldPersisted = InMemoryWorldStateProxy.persistState(worldToPersist) //State root hash needs to be up-to-date for prepared block
-    } yield BlockPreparationResult(execResult, worldPersisted.stateRootHash)
-
-    blockExecResult.foreach { _ => log.debug(s"Prepared for mining block with number ${block.header.number}") }
-
-    blockExecResult
+    } yield BlockPreparationResult(block, execResult, worldPersisted.stateRootHash)
+
+    blockExecResult match {
+      case Right(result) =>
+        log.debug(s"Prepared for mining block with number ${block.header.number}")
+        Right(result)
+      case Left(TxsExecutionError(tx,reason)) =>
+        log.debug(s"failed to execute transaction $tx because: $reason")
+        val newBlock = block.copy(body = block.body.copy(transactionList = block.body.transactionList.filter(_.hash != tx.hash)))
```

Comment:
1. yes we will re run `prepareBlock` recursively until we have no failing transactions (that may mean we end up with empty block)
2. No, it will be eventually pushed out of transaction pool by new incoming transactions, because we prepend them to the pool and we have finite pool size, so in some cases (not to many transactions) it may happen that we will keep this transaction forever